ratis-netty/src/main/java/org/apache/ratis/netty/server/NettyServerStreamRpc.java

```diff
@@ -123,11 +125,13 @@ void close() {
     private final List<DataStreamOutput> outs;
     private final AtomicReference<CompletableFuture<?>> previous
         = new AtomicReference<>(CompletableFuture.completedFuture(null));
+    private final ExecutorService writeExecutor;
```

Comment:
We should have a single executor for all the streams.  Otherwise, it requires a thread per stream.  Since threads are limited resource, the number of streams will be limited by the threads.  Also, creating more threads is not useful since io is bounded by the hardware (e.g. # of discs).

HDFS Datanode has the problem that it needs a thread per pipeline so that the number of pipeline is limited by it.

Reply:
@szetszwo Thanks the suggestions. I agree. Besides, I am not sure one thread is enough for all streams. So I create 4 ExecutorService, each has a single thread in the pool, and assign one of the 4  ExecutorService with the fewest tasks to the new stream. What do you think ?

Reply:
> ... I am not sure one thread is enough for all streams. ...

I mean a single executor with multiple threads.